llvm/lib/Target/DirectX/DXContainerGlobals.cpp

```diff
@@ -104,6 +112,38 @@ GlobalVariable *DXContainerGlobals::computeShaderHash(Module &M) {
   return GV;
 }
 
+GlobalVariable *DXContainerGlobals::buildContainerGlobal(
+    Module &M, Constant *Content, StringRef Name, StringRef SectionName) {
+  auto *GV = new llvm::GlobalVariable(
+      M, Content->getType(), true, GlobalValue::PrivateLinkage, Content, Name);
+  GV->setSection(SectionName);
+  GV->setAlignment(Align(4));
+  return GV;
+}
+
+GlobalVariable *DXContainerGlobals::buildSingature(Module &M, Signature &Sig,
+                                                   StringRef Name,
+                                                   StringRef SectionName) {
+  SmallString<256> Data;
+  raw_svector_ostream OS(Data);
+  Sig.write(OS);
+  Constant *Constant =
+      ConstantDataArray::getString(M.getContext(), Data, /*AddNull*/ false);
+  return buildContainerGlobal(M, Constant, Name, SectionName);
+}
+
+void DXContainerGlobals::addSingature(Module &M,
+                                      SmallVector<GlobalValue *> &Globals,
+                                      Triple &TT) {
+  Signature InputSig;
+  Signature OutputSig;
+  // FIXME: support graphics shader.
+  //  see issue https://github.com/llvm/llvm-project/issues/90504.
+
+  Globals.emplace_back(buildSingature(M, InputSig, "dx.isg1", "ISG1"));
+  Globals.emplace_back(buildSingature(M, OutputSig, "dx.osg1", "OSG1"));
```

Comment:
nit: It's clearer to read what is going on here if we group creating the signature with emitting it.
```suggestion
  // FIXME: support graphics shader.
  //  see issue https://github.com/llvm/llvm-project/issues/90504.
  Signature InputSig;
  Globals.emplace_back(buildSingature(M, InputSig, "dx.isg1", "ISG1"));

  Signature OutputSig;
  Globals.emplace_back(buildSingature(M, OutputSig, "dx.osg1", "OSG1"));
```

Reply:
Updated.